src/main/java/ttt/GameLoop.java

```diff
@@ -31,37 +31,54 @@ private void playGame() {
         while (!gameEngine.isOver()) {
             String input = getLocation();
             display.clearScreen();
-            if (isInvalidInput(input)) {
-                write(consoleBoard.createBoard(gameEngine.showBoard()));
-                display.invalidInput();
-            } else {
-                placeMark(input);
-            }
+            processInput(input);
         }
     }
 
+    private String getLocation() {
+        write(consoleBoard.createBoard(gameEngine.showBoard()));
+        display.displayTurn(gameEngine.currentMark());
+        display.promptForLocation();
+        return gameEngine.getCurrentPlayer().getLocation(input, gameEngine.showBoard());
+    }
+
     private void placeMark(String input) {
         int choice = Integer.parseInt(input);
         gameEngine.play(choice);
         write(consoleBoard.createBoard(gameEngine.showBoard()));
     }
 
+    private void processInput(String input) {
+        if (isInvalidInput(input)) {
+            display.invalidInput();
+        } else {
+            placeMark(input);
+        }
+    }
+
     private boolean isInvalidInput(String choice) {
         int location = Integer.parseInt(choice);
         if (location < 0 || location > 8) {
             display.invalidLocation();
             return true;
-        } else if (gameEngine.showBoard().getMarkAt(location) != Marks.CLEAR) {
+        } else if (gameEngine.board(location) != Marks.CLEAR) {
             display.takenCell();
             return true;
         }
         return false;
     }
 
-    private String getLocation() {
-        display.displayTurn(gameEngine.currentMark());
-        display.promptForLocation();
-        return gameEngine.getCurrentPlayer().getLocation(input, gameEngine.showBoard());
+    private void replay() {
+        display.replay();
+        playAgain(input.get());
+        display.goodbye();
+    }
+
+    private void playAgain(String answer) {
+        if (answer.equals("1")) {
```

Comment:
If as a client I say "I want to change the option from "1" to "R" for replay. how many places would change? Ideally it would be good if it was only one production file.